main.tf

```diff
@@ -254,4 +254,14 @@ resource "aws_lambda_event_source_mapping" "this" {
       uri  = source_access_configuration.value["uri"]
     }
   }
+
+  dynamic "filter_criteria" {
+    for_each = lookup(each.value, "filter_criteria", null) != null ? [true] : []
+
+    content {
+      filter {
+        pattern = each.value["filter_criteria"].pattern
```

Comment:
`pattern` is optional (according to the [docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping#filter_criteria-configuration-block)), so it should be using `lookup()`, too.